### ISSUE TITLE
async instance methods broken in 6.8.0

### ISSUE THREAD

**Issue by selbekk** [yannickcr/eslint-plugin-react#991]
We have a few async functions as instance methods in our component classes. An example is:

```javascript
class SessionHandler extends Component {
  async ping() {
    try {
      await server.ping();
    } catch(e) { this.props.onLoggedOut(); }
  }
  render() {
     // Render some stuff
  }
}
```

In #989 you introduced a check for whether a component was async or not. This broke code like this for us - which it shouldn't. Suggest that #989 is removed or rewritten to support this kind of syntax. Worked perfectly on 6.7.1 :)

**Comment by ljharb** [comment]
Can you elaborate on how it breaks? That change should only have affected component detection on functions.

**Comment by selbekk** [comment]
Oh sorry, of course. In the async method, the linter now think it is its own component. Therefore, I get linting errors like these:

```
'selectedCoverage' is missing in props validation  react/prop-types
'quantity' is missing in props validation          react/prop-types
```

on code like this:

```javascript
class SomeClass extends Component {
  async onAddToCart({ selectedCoverage, quantity }) {
    // ...
  }

  render() {
    return <CartWhatever onAddToCart={this.onAddToCart} />
  }
```

**Comment by ljharb** [comment]
Thanks, that definitely seems like a bug. @taion mind submitting a PR to fix this?

**Comment by taion** [comment]
Can't reproduce. See test cases added in #992.

**Comment by taion** [comment]
And in fact the PR should be expected to do the exact opposite, since the only added cases blacklist component detection.

**Comment by selbekk** [comment]
Thanks for the quick feedback guys :) I'll see if I can reproduce this outside of my application.

**Comment by selbekk** [comment]
Seems I can't reproduce this either outside of my own code. I'll close this until I have something to contribute.

**Comment by bharley** [comment]
I'm having the same issue:

```js
class CertSettings extends React.Component {
  static propTypes = {
    onSave: React.PropTypes.func.isRequired,
    errors: React.PropTypes.array,
  };

  handleSubmit = async ({ certificate: certificates, key: keys }) => {
    if (!certificates || !certificates.length || !keys || !keys.length) {
      return;
    }

    const { onSave } = this.props;
    await onSave(crtContents, keyContents);
  };

  // ...
}
```

Results in the following linting errors:

```
'certificate' is missing in props validation  react/prop-types
'key' is missing in props validation          react/prop-types
```

**Comment by ljharb** [comment]
@bharley does it have the same problem if you don't destructure in the signature?

(on a side note, there's zero reason you need an async function there, since you only have one `await` and you're not returning anything)

**Comment by bharley** [comment]
@ljharb: It does not have a problem with that, no.

I trimmed down the function to show the relevant parts. It does look silly though...

You know what, that got me thinking... I played around with the code to it reproduce. This has issues:

```js
export class Example extends React.Component {
  static propTypes = {
    certificates: React.PropTypes.array,
    onSave: React.PropTypes.func.isRequired,
    onDelete: React.PropTypes.func.isRequired,
    deleteFingerprint: React.PropTypes.string.isRequired,
  };

  handleDeleteConfirm = () => {
    const { deleteFingerprint } = this.props;
    this.props.onDelete(deleteFingerprint);
  };

  handleSubmit = async ({ certificate: certificates, key: keys }) => {
    if (!certificates || !certificates.length || !keys || !keys.length) {
      return;
    }

    const { onSave } = this.props;
    await onSave(certificates, keys);
  };
}
```

but this doesn't

```js
export class Example extends React.Component {
  static propTypes = {
    certificates: React.PropTypes.array,
    onSave: React.PropTypes.func.isRequired,
    onDelete: React.PropTypes.func.isRequired,
    deleteFingerprint: React.PropTypes.string.isRequired,
  };

  handleSubmit = async ({ certificate: certificates, key: keys }) => {
    if (!certificates || !certificates.length || !keys || !keys.length) {
      return;
    }

    const { onSave } = this.props;
    await onSave(certificates, keys);
  };
}
```
